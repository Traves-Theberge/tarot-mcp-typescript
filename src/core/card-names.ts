/**
 * Utility functions for generating tarot card names
 */

import { MajorArcana, CardValue, Suit, TarotCard } from './tarot-card.js';

/**
 * Get the display name for a Major Arcana card
 */
export function getMajorArcanaName(arcana: MajorArcana): string {
  switch (arcana) {
    case MajorArcana.Fool:
      return 'The Fool';
    case MajorArcana.Magician:
      return 'The Magician';
    case MajorArcana.HighPriestess:
      return 'The High Priestess';
    case MajorArcana.Empress:
      return 'The Empress';
    case MajorArcana.Emperor:
      return 'The Emperor';
    case MajorArcana.Hierophant:
      return 'The Hierophant';
    case MajorArcana.Lovers:
      return 'The Lovers';
    case MajorArcana.Chariot:
      return 'The Chariot';
    case MajorArcana.Strength:
      return 'Strength';
    case MajorArcana.Hermit:
      return 'The Hermit';
    case MajorArcana.WheelOfFortune:
      return 'Wheel of Fortune';
    case MajorArcana.Justice:
      return 'Justice';
    case MajorArcana.HangedMan:
      return 'The Hanged Man';
    case MajorArcana.Death:
      return 'Death';
    case MajorArcana.Temperance:
      return 'Temperance';
    case MajorArcana.Devil:
      return 'The Devil';
    case MajorArcana.Tower:
      return 'The Tower';
    case MajorArcana.Star:
      return 'The Star';
    case MajorArcana.Moon:
      return 'The Moon';
    case MajorArcana.Sun:
      return 'The Sun';
    case MajorArcana.Judgement:
      return 'Judgement';
    case MajorArcana.World:
      return 'The World';
    default:
      throw new Error(`Unknown Major Arcana: ${arcana}`);
  }
}

/**
 * Get the display name for a card value
 */
export function getCardValueName(value: CardValue): string {
  switch (value) {
    case CardValue.Ace:
      return 'Ace';
    case CardValue.Two:
      return 'Two';
    case CardValue.Three:
      return 'Three';
    case CardValue.Four:
      return 'Four';
    case CardValue.Five:
      return 'Five';
    case CardValue.Six:
      return 'Six';
    case CardValue.Seven:
      return 'Seven';
    case CardValue.Eight:
      return 'Eight';
    case CardValue.Nine:
      return 'Nine';
    case CardValue.Ten:
      return 'Ten';
    case CardValue.Page:
      return 'Page';
    case CardValue.Knight:
      return 'Knight';
    case CardValue.Queen:
      return 'Queen';
    case CardValue.King:
      return 'King';
    default:
      throw new Error(`Unknown Card Value: ${value}`);
  }
}

/**
 * Get the display name for a Minor Arcana card
 */
export function getMinorArcanaName(suit: Suit, value: CardValue): string {
  const suitName = suit.charAt(0).toUpperCase() + suit.slice(1);
  return `${getCardValueName(value)} of ${suitName}`;
}

/**
 * Get the display name for any tarot card
 */
export function getCardName(card: TarotCard): string {
  switch (card.type) {
    case 'major':
      return getMajorArcanaName(card.arcana as MajorArcana);
    case 'minor':
      const minorArcana = card.arcana as { suit: Suit; value: CardValue };
      return getMinorArcanaName(minorArcana.suit, minorArcana.value);
    default:
      throw new Error(`Unknown card type: ${(card as any).type}`);
  }
} 