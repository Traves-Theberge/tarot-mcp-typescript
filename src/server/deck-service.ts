/**
 * Deck service for drawing and managing tarot cards
 */

import { TarotCard, TarotCardError, createFullDeck, DECK_CONSTANTS } from '../core/index.js';
import {
  RandomNumberGenerator,
  SystemRandomNumberGenerator,
  shuffleArray,
} from '../utils/index.js';

/**
 * Service class for managing tarot deck operations
 */
export class TarotDeckService {
  private static readonly FULL_DECK = createFullDeck();
  private static readonly SHUFFLE_ITERATIONS = 10;

  constructor(private readonly rng: RandomNumberGenerator = new SystemRandomNumberGenerator()) {}

  /**
   * Draw cards from the deck
   * 
   * @param count - Number of cards to draw (1-78)
   * @returns Array of drawn cards
   * @throws TarotCardError if count is invalid
   */
  drawCards(count: number): readonly TarotCard[] {
    this.validateCardCount(count);

    // Create a copy of the full deck
    let deck = [...TarotDeckService.FULL_DECK];

    // Shuffle the deck multiple times for better randomness
    for (let i = 0; i < TarotDeckService.SHUFFLE_ITERATIONS; i++) {
      deck = shuffleArray(deck, this.rng);
    }

    // Return the requested number of cards
    return deck.slice(0, count);
  }

  /**
   * Get the complete tarot deck (78 cards)
   * 
   * @returns Array of all tarot cards
   */
  getFullDeck(): readonly TarotCard[] {
    return [...TarotDeckService.FULL_DECK];
  }

  /**
   * Get the total number of cards in a complete deck
   */
  get deckSize(): number {
    return DECK_CONSTANTS.TOTAL_CARDS;
  }

  /**
   * Validate card count is within valid range
   */
  private validateCardCount(count: number): void {
    if (count < 1 || count > DECK_CONSTANTS.TOTAL_CARDS) {
      throw new TarotCardError(`Invalid card count: ${count}. Must be between 1 and ${DECK_CONSTANTS.TOTAL_CARDS}.`);
    }
  }
} 