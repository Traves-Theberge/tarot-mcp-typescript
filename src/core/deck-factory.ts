/**
 * Factory functions for creating tarot decks and collections
 */

import { MajorArcana, Suit, CardValue, TarotCard } from './tarot-card.js';
import { getCardName } from './card-names.js';

/**
 * Create all Major Arcana cards
 */
export function createMajorArcanaCards(): readonly TarotCard[] {
  return Object.values(MajorArcana)
    .filter((value): value is MajorArcana => typeof value === 'number')
    .map(
      (arcana): TarotCard => {
        const tempCard: TarotCard = {
          type: 'major',
          arcana,
          name: '',
          uri: '',
          imagePath: ''
        };
        const name = getCardName(tempCard);
        const uri = `tarot://card/major/${arcana}`;
        const imagePath = `assets/images/major_arcana_${name.toLowerCase().replace(/\s+/g, '_')}.png`;
        
        return {
          type: 'major',
          arcana,
          name,
          uri,
          imagePath
        };
      }
    );
}

/**
 * Create all Minor Arcana cards
 */
export function createMinorArcanaCards(): readonly TarotCard[] {
  const cards: TarotCard[] = [];

  for (const suit of Object.values(Suit)) {
    for (const value of Object.values(CardValue)) {
      if (typeof value === 'string') {
        const arcana = { suit, value };
        const tempCard: TarotCard = {
          type: 'minor',
          arcana,
          name: '',
          uri: '',
          imagePath: ''
        };
        const name = getCardName(tempCard);
        const uri = `tarot://card/minor/${suit}/${value}`;
        const imagePath = `assets/images/minor_arcana_${suit}_${value}.png`;
        
        cards.push({
          type: 'minor',
          arcana,
          name,
          uri,
          imagePath
        });
      }
    }
  }

  return cards;
}

/**
 * Create the complete tarot deck (78 cards)
 */
export function createFullDeck(): readonly TarotCard[] {
  return [...createMajorArcanaCards(), ...createMinorArcanaCards()];
}

/**
 * Deck constants
 */
export const DECK_CONSTANTS = {
  MAJOR_ARCANA_COUNT: 22,
  MINOR_ARCANA_COUNT: 56,
  TOTAL_CARDS: 78,
  SUITS_COUNT: 4,
  VALUES_PER_SUIT: 14,
} as const; 