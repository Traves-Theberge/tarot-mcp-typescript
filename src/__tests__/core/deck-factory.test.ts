/**
 * Tests for TarotDeck functionality
 * TypeScript equivalent of Swift TarotDeckTests
 */

import { MajorArcana, Suit, CardValue, TarotCard } from '../../core/index.js';
import { createFullDeck } from '../../core/deck-factory.js';
import { getCardName } from '../../core/card-names.js';
import { TarotDeckService } from '../../server/deck-service.js';
import { CountedSet, SeedablePseudoRNG } from '../test-support/index.js';

describe('Tarot Deck Tests', () => {
  const fullDeck = createFullDeck();

  describe('Full Deck', () => {
    test('has exactly 78 cards', () => {
      expect(fullDeck).toHaveLength(78);
    });

    test('contains all 22 Major Arcana cards', () => {
      const majorCards = fullDeck.filter(card => card.type === 'major');
      expect(majorCards).toHaveLength(22);

      // Check that all Major Arcana are present
      const majorArcanaValues = Object.values(MajorArcana).filter(v => typeof v === 'number') as number[];
      for (const expectedArcana of majorArcanaValues) {
        const found = majorCards.some(card => 
          card.type === 'major' && card.arcana === expectedArcana
        );
        expect(found).toBe(true);
      }
    });

    test('contains all 56 Minor Arcana cards', () => {
      const minorCards = fullDeck.filter(card => card.type === 'minor');
      expect(minorCards).toHaveLength(56);

      const suits = Object.values(Suit);
      const values = Object.values(CardValue).filter(v => typeof v === 'number') as number[];
      
      expect(suits).toHaveLength(4);
      expect(values).toHaveLength(14);

      // Check that all combinations of suits and values are present
      for (const suit of suits) {
        for (const value of values) {
          const found = minorCards.some(card =>
            card.type === 'minor' && 
            card.arcana.suit === suit && 
            card.arcana.value === value
          );
          expect(found).toBe(true);
        }
      }
    });
  });

  describe('Draw Cards', () => {
    // Repeat test multiple times like Swift's .repeat(count: 10)
    test.each(Array.from({ length: 10 }, (_, i) => i))('drawCards with count=1 returns a valid card from the deck (iteration %i)', () => {
      const rng = new SeedablePseudoRNG(Math.random() * 1000000);
      const deckService = new TarotDeckService(rng);
      const cards = deckService.drawCards(1);
      expect(cards).toHaveLength(1);
      
      const cardNames = fullDeck.map(getCardName);
      expect(cards.length).toBeGreaterThan(0);
      if (cards.length > 0) {
        expect(cardNames).toContain(getCardName(cards[0]!));
      }
    });

    test.each(Array.from({ length: 10 }, (_, i) => i))('drawCards returns correct number of cards (iteration %i)', () => {
      const rng = new SeedablePseudoRNG(Math.random() * 1000000);
      const deckService = new TarotDeckService(rng);
      
      const oneCard = deckService.drawCards(1);
      expect(oneCard).toHaveLength(1);

      const threeCards = deckService.drawCards(3);
      expect(threeCards).toHaveLength(3);

      const tenCards = deckService.drawCards(10);
      expect(tenCards).toHaveLength(10);
    });

    test.each(Array.from({ length: 10 }, (_, i) => i))('drawCards returns valid cards from the deck (iteration %i)', () => {
      const rng = new SeedablePseudoRNG(Math.random() * 1000000);
      const deckService = new TarotDeckService(rng);
      const cards = deckService.drawCards(5);

      const fullDeckNames = new Set(fullDeck.map(getCardName));
      for (const card of cards) {
        expect(fullDeckNames.has(getCardName(card))).toBe(true);
      }
    });

    test('drawCards with count larger than deck throws error', () => {
      const rng = new SeedablePseudoRNG(42);
      const deckService = new TarotDeckService(rng);
      expect(() => deckService.drawCards(100)).toThrow('Invalid card count: 100. Count must be between 1 and 78.');
    });

    test('drawCards with zero count throws error', () => {
      const rng = new SeedablePseudoRNG(42);
      const deckService = new TarotDeckService(rng);
      expect(() => deckService.drawCards(0)).toThrow();
    });

    test('drawRandomCard with seeded RNG is deterministic', () => {
      const rng1 = new SeedablePseudoRNG(42);
      const rng2 = new SeedablePseudoRNG(42);
      const deckService1 = new TarotDeckService(rng1);
      const deckService2 = new TarotDeckService(rng2);

      const cards1 = deckService1.drawCards(1);
      const cards2 = deckService2.drawCards(1);

      expect(cards1.length).toBe(1);
      expect(cards2.length).toBe(1);
      if (cards1.length > 0 && cards2.length > 0) {
        expect(getCardName(cards1[0]!)).toBe(getCardName(cards2[0]!));
      }
    });

    test.each(Array.from({ length: 10 }, (_, i) => i))('drawCards with seeded RNG is deterministic (iteration %i)', () => {
      const seed = Math.floor(Math.random() * 1000000);
      const rng1 = new SeedablePseudoRNG(seed);
      const rng2 = new SeedablePseudoRNG(seed);
      const deckService1 = new TarotDeckService(rng1);
      const deckService2 = new TarotDeckService(rng2);

      const cards1 = deckService1.drawCards(5);
      const cards2 = deckService2.drawCards(5);

      expect(cards1).toHaveLength(cards2.length);

      for (let i = 0; i < cards1.length; i++) {
        expect(getCardName(cards1[i]!)).toBe(getCardName(cards2[i]!));
      }
    });

    test.each(Array.from({ length: 10 }, (_, i) => i))('Seeded RNG produces consistent sequences (iteration %i)', () => {
      const seed = Math.floor(Math.random() * 1000000);
      const rng1 = new SeedablePseudoRNG(seed);
      const rng2 = new SeedablePseudoRNG(seed);
      const deckService1 = new TarotDeckService(rng1);
      const deckService2 = new TarotDeckService(rng2);

      const sequence1 = Array.from({ length: 10 }, () => deckService1.drawCards(1)[0]);
      const sequence2 = Array.from({ length: 10 }, () => deckService2.drawCards(1)[0]);

      for (let i = 0; i < sequence1.length; i++) {
        expect(getCardName(sequence1[i]!)).toBe(getCardName(sequence2[i]!));
      }
    });

    test('drawCards with seeded RNG snapshot', () => {
      const rng = new SeedablePseudoRNG(38474);
      const deckService = new TarotDeckService(rng);
      const cards = deckService.drawCards(5);
      
      // Convert to a format similar to Swift's snapshot
      const cardNames = cards.map(getCardName);
      expect(cardNames).toMatchSnapshot();
    });

    test('Consecutive single card draws are evenly distributed', () => {
      const rng = new SeedablePseudoRNG(12345);
      const deckService = new TarotDeckService(rng);
      const picks = new CountedSet<string>();
      
      // Draw 78,000 cards like in Swift test
      for (let i = 0; i < 78000; i++) {
        const cards = deckService.drawCards(1);
        expect(cards.length).toBe(1);
        if (cards.length > 0) {
          picks.insert(getCardName(cards[0]!));
        }
      }

      // Each card should appear roughly 1000 times (78000 / 78 = 1000)
      // Allow range of 500-1500 like Swift test
      for (const card of fullDeck) {
        const count = picks.count(getCardName(card));
        expect(count).toBeGreaterThanOrEqual(500);
        expect(count).toBeLessThanOrEqual(1500);
      }
    });
  });
}); 