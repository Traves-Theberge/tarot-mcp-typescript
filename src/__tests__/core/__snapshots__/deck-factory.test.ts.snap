// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tarot Deck Tests Draw Cards drawCards with seeded RNG snapshot 1`] = `
[
  "Queen of Swords",
  "The Lovers",
  "Eight of Pentacles",
  "The Hierophant",
  "Ten of Pentacles",
]
`;
